examples/rust/tracecon/src/bpf/tracecon.bpf.c

```diff
@@ -0,0 +1,162 @@
+// SPDX-License-Identifier: GPL-2.0 OR BSD-3-Clause
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+
+#define memcpy(dest, src, n) __builtin_memcpy((dest), (src), (n))
+#define HOSTNAME_LEN 84
+
+const volatile pid_t target_pid = 0;
+
+/* Copied from: include/netdb.h */
+struct addrinfo {
+	int ai_flags; /* Input flags.  */
+	int ai_family; /* Protocol family for socket.  */
+	int ai_socktype; /* Socket type.  */
+	int ai_protocol; /* Protocol for socket.  */
+	u32 ai_addrlen; /* Length of socket address.  */
+	struct sockaddr *ai_addr; /* Socket address for socket.  */
+	char *ai_canonname; /* Canonical name for service location.  */
+	struct addrinfo *ai_next; /* Pointer to next in list.  */
+};
+
+struct lookup {
+	char c[84];
+	struct addrinfo **results;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 10240);
+	__type(key, u32);
+	__type(value, struct lookup);
+} lookups SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 10240);
+	__type(key, u32);
+	__type(value, struct lookup);
+} hostnames SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 10240);
+	__type(key, u32);
+	__type(value, struct sock *);
+} sockets SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+} events SEC(".maps");
+
+enum tag { IP = 0, HOSTNAME = 1 };
+
+struct event {
+	u8 tag;
+	u8 ip[4];
+	u8 hostname[HOSTNAME_LEN];
+};
+
+struct event _event = {};
+
+static u32 get_tid()
+{
+	u64 tgid = bpf_get_current_pid_tgid();
+	pid_t pid = tgid >> 32;
+
+	if (target_pid != 0 && target_pid != pid)
+		return 0;
+	return (u32)tgid;
+}
+
+SEC("uprobe/getaddrinfo")
+int BPF_KPROBE(getaddrinfo_enter, const char *hostname, const char *service,
+	       const struct addrinfo *hints, struct addrinfo **res)
+{
+	u32 tid = get_tid();
+	struct lookup lookup = {};
+
+	if (!tid)
+		return 0;
+	bpf_probe_read_user_str(&lookup.c, sizeof(lookup.c), hostname);
+	lookup.results = res;
+	bpf_map_update_elem(&lookups, &tid, &lookup, BPF_ANY);
+	return 0;
+}
+
+SEC("uretprobe/getaddrinfo")
+int BPF_KRETPROBE(getaddrinfo_exit, int ret)
+{
+	u32 tid = get_tid();
+	struct lookup *lookup;
+	struct addrinfo *result;
+	struct sockaddr *addr;
+	struct in_addr ipv4_addr;
+
+	if (!tid)
+		return 0;
+	if (ret != 0)
+		goto cleanup;
+	lookup = bpf_map_lookup_elem(&lookups, &tid);
+	if (!lookup)
+		return 0;
+	bpf_probe_read_user(&result, sizeof(result), lookup->results);
+	bpf_probe_read_user(&addr, sizeof(addr), &result->ai_addr);
+	bpf_probe_read_user(&ipv4_addr, sizeof(ipv4_addr),
+	                    &((struct sockaddr_in *)addr)->sin_addr);
```

Comment:
Ok, sure, you can also declare `struct sockaddr_in *addr` instead of `struct sockaddr *addr;`, then you'll avoid extra cast.